### PR TITLE
Snap 2351

### PR DESCRIPTION
**Pull request by vivekwiz**
## Changes proposed in this pull request
From lead node send back metadata for prepared statement  at prepare phase too.

## Patch testing
Local testing

TODO:
More unit tests
Prepcheckin

## ReleaseNotes.txt changes
No

## Other PRs 
https://github.com/SnappyDataInc/snappy-store/pull/387

### CODE REVIEW

File: cluster/src/main/scala/io/snappydata/gemxd/SparkSQLPrepareImpl.scala [hunk 1]
```diff
@@ -225,6 +241,22 @@ class SparkSQLPrepareImpl(val sql: String,
   }
 }
 
+object SparkSQLPrepareImpl{
+
+  def getTableNamesAndDatatype(output: Seq[expressions.Attribute]):
+  (Array[String], Array[DataType]) = {
+    var i = 0
```
Comment (rishitesh) [inline cluster/src/main/scala/io/snappydata/gemxd/SparkSQLPrepareImpl.scala]: Can be a simple map function which can remove the var i;

File: cluster/src/main/scala/io/snappydata/gemxd/SparkSQLExecuteImpl.scala [hunk 2]
```diff
@@ -302,9 +273,54 @@ class SparkSQLExecuteImpl(val sql: String,
...
   }
-}
 
-object SparkSQLExecuteImpl {
+  def getTableNamesAndNullability(output: Seq[expressions.Attribute]):
+  (Array[String], Array[Boolean]) = {
+    var i = 0
+    val tables = new Array[String](output.length)
+    val nullables = new Array[Boolean](output.length)
+    output.foreach { a =>
```
Comment (rishitesh) [inline cluster/src/main/scala/io/snappydata/gemxd/SparkSQLExecuteImpl.scala]: Map function can be used here.

File: cluster/src/test/scala/io/snappydata/cluster/PreparedQueryRoutingSingleNodeSuite.scala [hunk 3]
```diff
@@ -241,6 +241,33 @@ class PreparedQueryRoutingSingleNodeSuite extends SnappyFunSuite with BeforeAndA
...
+    SnappySession.getPlanCache.invalidateAll()
+    assert(SnappySession.getPlanCache.asMap().size() == 0)
+    SnappyTableStatsProviderService.suspendCacheInvalidation = true
+    val tableName = "order_line_col"
+    try {
+      snc.sql(s"create table $tableName (ol_int_id  integer," +
+          s" ol_int2_id  integer, ol_str_id STRING) using column " +
+          "options( partition_by 'ol_int_id, ol_int2_id', buckets '2')")
+
+      val serverHostPort = TestUtil.startNetServer()
```
Comment (rishitesh) [inline cluster/src/test/scala/io/snappydata/cluster/PreparedQueryRoutingSingleNodeSuite.scala]: in finally we should stop the net server.

File: cluster/src/main/scala/io/snappydata/gemxd/SparkSQLPrepareImpl.scala (hunk 1 above)
Comment (kneeraj) [reply cluster/src/main/scala/io/snappydata/gemxd/SparkSQLPrepareImpl.scala]: Yes. Let me change it. I just focused on correctness in Vivek's changes.

File: cluster/src/main/scala/io/snappydata/gemxd/SparkSQLExecuteImpl.scala (hunk 2 above)
Comment (kneeraj) [reply cluster/src/main/scala/io/snappydata/gemxd/SparkSQLExecuteImpl.scala]: Yep.

**Pull request merged**